luminoso_api/v5_client.py

```diff
@@ -0,0 +1,372 @@
+        return self._json_request('get', url, params=params)
+
+    def post(self, path='', **params):
+        """
+        Make a POST request to the given path, and return the JSON-decoded
+        result.
+
+        Keyword parameters will be converted to form values, sent in the body
+        of the POST.
+
+        POST requests are requests that cause a change on the server,
+        especially those that ask to create and return an object of some kind.
+        """
+        url = ensure_trailing_slash(self.url + path.lstrip('/'))
+        return self._json_request('post', url, data=json.dumps(params),
+                                  headers={'Content-Type': 'application/json'})
+
+    def put(self, path='', **params):
+        """
+        Make a PUT request to the given path, and return the JSON-decoded
+        result.
+
+        Keyword parameters will be converted to form values, sent in the body
+        of the PUT.
+
+        PUT requests are usually requests to *update* the object represented by
+        this URL. Unlike POST requests, PUT requests can be safely duplicated.
+        """
+        url = ensure_trailing_slash(self.url + path.lstrip('/'))
+        return self._json_request('put', url, data=json.dumps(params),
+                                  headers={'Content-Type': 'application/json'})
+
+    def patch(self, path='', **params):
+        """
+        Make a PATCH request to the given path, and return the JSON-decoded
+        result.
+
+        Keyword parameters will be converted to form values, sent in the body
+        of the PATCH.
+
+        PATCH requests are usually requests to make *small fixes* to the
+        object represented by this URL.
+        """
+        url = ensure_trailing_slash(self.url + path.lstrip('/'))
+        return self._json_request('patch', url, data=json.dumps(params),
+                                  headers={'Content-Type': 'application/json'})
+
+    def delete(self, path='', **params):
+        """
+        Make a DELETE request to the given path, and return the JSON-decoded
+        result.
+
+        Keyword parameters will be converted to URL parameters.
+
+        DELETE requests ask to delete the object represented by this URL.
+        """
+        params = jsonify_parameters(params)
+        url = ensure_trailing_slash(self.url + path.lstrip('/'))
+        return self._json_request('delete', url, params=params)
+
+    # Useful abstractions
+    def change_path(self, path):
+        """
+        Change the client's path. Paths with leading slashes are appended to
+        the root url, otherwise paths are set relative to the current path.
+        Returns a copy of the client to avoid breaking old code.
+        """
+        if path.startswith('/'):
+            self.url = ensure_trailing_slash(self.root_url + path)
+        else:
+            self.url = ensure_trailing_slash(self.url + path)
+        return self
+
+    def upload(self, path, docs, **params):
```

Comment:
Given that we're now looking at releasing this with the v4 and v5 clients available side-by-side, I doubt there's much advantage to keeping this method around -- people might as well switch to a regular post as they do the rest of the migration to v5. Unless you think this is likely to give people useful feedback while converting their scripts? My suspicion is that there are enough changes that using upload() vs post() will be the least of their worries, but it wouldn't take much argument to change my mind.

Reply:
I was thinking of this not as part of scripts but as part of interactive use: that if people are uploading things to the API by, as I do, going into Python and instantiating a client and calling its methods, that it would be easy to call `.upload()` by habit.  In which case I wanted a warning rather than just an AttributeError.

I could see removing this; or maybe changing it to a NotImplementedError.  While I think you're right that "there are enough changes" in the v4 vs. v5 API, those changes are almost all "use a different endpoint as the argument" or "you get back a different result", and this is roughly the only case where it's an actual fact about the *client* that's different.